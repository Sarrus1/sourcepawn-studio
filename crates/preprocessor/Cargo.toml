[package]
name = "preprocessor"
version = "0.1.0"
description = "SourcePawn preprocessor written in Rust"
include = ["Cargo.toml", "src/"]
edition.workspace = true
license.workspace = true
authors.workspace = true

[lib]
doctest = false

[dependencies]
lsp-types.workspace = true
anyhow.workspace = true
lazy_static.workspace = true
regex.workspace = true
fxhash.workspace = true
sourcepawn_lexer.workspace = true
salsa.workspace = true
log.workspace = true
smol_str.workspace = true
itertools.workspace = true
deepsize.workspace = true
serde.workspace = true
la-arena.workspace = true

vfs.workspace = true
base-db.workspace = true
stdx.workspace = true
syntax.workspace = true

[dev-dependencies]
criterion = "0.5.1"
minreq = { version = "2.7.0", features = ["https"] }
insta.workspace = true

[[bench]]
name = "bench_main"
harness = false
path = "benches/bench_main.rs"
