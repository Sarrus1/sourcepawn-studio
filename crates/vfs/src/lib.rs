//! # Virtual File System
//!
//! **Note:** This is heavily inspired from rust-analyzer.
//!
//! VFS stores all files read by sourcepawn-studio. Reading file contents from VFS
//! always returns the same contents, unless VFS was explicitly modified with
//! [`set_file_contents`]. All changes to VFS are logged, and can be retrieved via
//! the [`take_changes`] method.
//!
//! Files in VFS are identified with [`FileId`]s -- interned paths.
//!
//! [`set_file_contents`]: Vfs::set_file_contents
//! [`take_changes`]: Vfs::take_changes

use std::{
    fmt::{self},
    mem,
};

use fxhash::FxHashMap;
use path_interner::PathInterner;

mod anchored_path;
mod file_id;
mod file_set;
pub mod loader;
mod path_interner;
mod vfs_path;

pub use {
    anchored_path::{AnchoredPath, AnchoredPathBuf},
    file_id::FileId,
    file_set::{FileSet, FileSetConfig},
    vfs_path::VfsPath,
};

/// Storage for all files read by rust-analyzer.
///
/// For more information see the [crate-level](crate) documentation.
#[derive(Default)]
pub struct Vfs {
    interner: PathInterner,
    data: Vec<Option<Vec<u8>>>,
    changes: Vec<ChangedFile>,
}

/// Changed file in the [`Vfs`].
#[derive(Debug)]
pub struct ChangedFile {
    /// Id of the changed file
    pub file_id: FileId,
    /// Kind of change
    pub change_kind: ChangeKind,
}

impl ChangedFile {
    /// Returns `true` if the change is not [`Delete`](ChangeKind::Delete).
    pub fn exists(&self) -> bool {
        self.change_kind != ChangeKind::Delete
    }

    /// Returns `true` if the change is [`Create`](ChangeKind::Create) or
    /// [`Delete`](ChangeKind::Delete).
    pub fn is_created_or_deleted(&self) -> bool {
        matches!(self.change_kind, ChangeKind::Create | ChangeKind::Delete)
    }
}

/// Kind of [file change](ChangedFile).
#[derive(Eq, PartialEq, Copy, Clone, Debug)]
pub enum ChangeKind {
    /// The file was (re-)created
    Create,
    /// The file was modified
    Modify,
    /// The file was deleted
    Delete,
}

impl Vfs {
    /// Id of the given path if it exists in the `Vfs` and is not deleted.
    pub fn file_id(&self, path: &VfsPath) -> Option<FileId> {
        self.interner.get(path).filter(|&it| self.get(it).is_some())
    }

    /// File path corresponding to the given `file_id`.
    ///
    /// # Panics
    ///
    /// Panics if the id is not present in the `Vfs`.
    pub fn file_path(&self, file_id: FileId) -> VfsPath {
        self.interner.lookup(file_id).clone()
    }

    /// File content corresponding to the given `file_id`.
    ///
    /// # Panics
    ///
    /// Panics if the id is not present in the `Vfs`, or if the corresponding file is
    /// deleted.
    pub fn file_contents(&self, file_id: FileId) -> &[u8] {
        self.get(file_id).as_deref().unwrap()
    }

    /// Returns the overall memory usage for the stored files.
    pub fn memory_usage(&self) -> usize {
        self.data.iter().flatten().map(|d| d.capacity()).sum()
    }

    /// Returns an iterator over the stored ids and their corresponding paths.
    ///
    /// This will skip deleted files.
    pub fn iter(&self) -> impl Iterator<Item = (FileId, &VfsPath)> + '_ {
        (0..self.data.len())
            .map(|it| FileId(it as u32))
            .filter(move |&file_id| self.get(file_id).is_some())
            .map(move |file_id| {
                let path = self.interner.lookup(file_id);
                (file_id, path)
            })
    }

    /// Update the `path` with the given `contents`. `None` means the file was deleted.
    ///
    /// Returns `true` if the file was modified, and saves the [change](ChangedFile).
    ///
    /// If the path does not currently exists in the `Vfs`, allocates a new
    /// [`FileId`] for it.
    pub fn set_file_contents(&mut self, path: VfsPath, mut contents: Option<Vec<u8>>) -> bool {
        let file_id = self.alloc_file_id(path);
        let change_kind = match (self.get(file_id), &contents) {
            (None, None) => return false,
            (Some(old), Some(new)) if old == new => return false,
            (None, Some(_)) => ChangeKind::Create,
            (Some(_), None) => ChangeKind::Delete,
            (Some(_), Some(_)) => ChangeKind::Modify,
        };
        if let Some(contents) = &mut contents {
            contents.shrink_to_fit();
        }
        *self.get_mut(file_id) = contents;
        self.changes.push(ChangedFile {
            file_id,
            change_kind,
        });
        true
    }

    /// Returns `true` if the `Vfs` contains [changes](ChangedFile).
    pub fn has_changes(&self) -> bool {
        !self.changes.is_empty()
    }

    /// Drain and returns all the changes in the `Vfs`.
    pub fn take_changes(&mut self) -> Vec<ChangedFile> {
        mem::take(&mut self.changes)
    }

    /// Provides a panic-less way to verify file_id validity.
    pub fn exists(&self, file_id: FileId) -> bool {
        self.get(file_id).is_some()
    }

    /// Returns the id associated with `path`
    ///
    /// - If `path` does not exists in the `Vfs`, allocate a new id for it, associated with a
    ///   deleted file;
    /// - Else, returns `path`'s id.
    ///
    /// Does not record a change.
    fn alloc_file_id(&mut self, path: VfsPath) -> FileId {
        let file_id = self.interner.intern(path);
        let idx = file_id.0 as usize;
        let len = self.data.len().max(idx + 1);
        self.data.resize_with(len, || None);
        file_id
    }

    /// Returns the content associated with the given `file_id`.
    ///
    /// # Panics
    ///
    /// Panics if no file is associated to that id.
    fn get(&self, file_id: FileId) -> &Option<Vec<u8>> {
        &self.data[file_id.0 as usize]
    }

    /// Mutably returns the content associated with the given `file_id`.
    ///
    /// # Panics
    ///
    /// Panics if no file is associated to that id.
    fn get_mut(&mut self, file_id: FileId) -> &mut Option<Vec<u8>> {
        &mut self.data[file_id.0 as usize]
    }

    /// Returns a map of all the files in the `Vfs`.
    pub fn get_url_map(&self) -> FxHashMap<FileId, VfsPath> {
        let mut res = FxHashMap::default();
        for (file_id, path) in self.iter() {
            res.insert(file_id, path.clone());
        }

        res
    }
}

impl fmt::Debug for Vfs {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Vfs")
            .field("n_files", &self.data.len())
            .finish()
    }
}
