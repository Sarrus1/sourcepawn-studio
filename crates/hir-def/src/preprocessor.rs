use std::{
    hash::{Hash, Hasher},
    sync::Arc,
};

use fxhash::FxHashMap;
use preprocessor::{Macro, PreprocessingResult, SourcepawnPreprocessor};
use vfs::{AnchoredPath, FileId};

use crate::DefDatabase;

// FIXME: Everything about this is bad.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct HashableHashMap<K: Hash + Eq, V: Eq> {
    pub map: FxHashMap<K, V>,
}

impl<K: Hash + Ord, V: Hash + Eq> Hash for HashableHashMap<K, V> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        // Collect and sort the keys to ensure consistent order
        let mut pairs: Vec<_> = self.map.iter().collect();
        pairs.sort_by(|a, b| a.0.cmp(b.0));

        // Hash each key-value pair
        for (key, value) in pairs {
            key.hash(state);
            value.hash(state);
        }
    }
}

pub(crate) fn preprocess_file_query(
    db: &dyn DefDatabase,
    file_id: FileId,
) -> Arc<PreprocessingResult> {
    // FIXME: Handle cyclic dependencies
    // FIXME: Handle errors
    let root_file_id = db.projet_subgraph(file_id).unwrap().root.file_id;
    let res = db.preprocess_file_inner(root_file_id, HashableHashMap::default());

    res.get(&file_id).unwrap().clone()
}

// FIXME: When preprocessing subfiles, we do not know the macros of the parent file.
// Solutions:
// 1. Make a struct that holds the macros that is always equal and always has the same hash?
// 2. Get all the parents's macros and pass them to the subfile's preprocessing function?
pub(crate) fn _preprocess_file_query(
    db: &dyn DefDatabase,
    file_id: FileId,
    macros: HashableHashMap<String, Macro>,
) -> Arc<FxHashMap<FileId, Arc<PreprocessingResult>>> {
    // FIXME: Handle cyclic dependencies
    let text = db.file_text(file_id);
    let mut results: FxHashMap<FileId, Arc<PreprocessingResult>> = FxHashMap::default();
    let mut preprocessor = SourcepawnPreprocessor::new(file_id, &text);
    preprocessor.set_macros(macros.map);
    let res = preprocessor
        .preprocess_input(
            &mut (|macros: &mut FxHashMap<String, Macro>,
                   path: String,
                   file_id: FileId,
                   quoted: bool| {
                let mut inc_file_id = None;
                if quoted {
                    inc_file_id = db.resolve_path(AnchoredPath::new(file_id, &path));
                };
                if inc_file_id.is_none() {
                    inc_file_id = db.resolve_path_relative_to_roots(&path);
                }
                if let Some(inc_file_id) = inc_file_id {
                    let map = HashableHashMap {
                        map: macros.clone(),
                    };
                    let res = db.preprocess_file_inner(inc_file_id, map);
                    results.extend(res.as_ref().clone()); // FIXME: This is bad
                    macros.extend(res.as_ref().get(&inc_file_id).unwrap().macros().clone());
                }

                Ok(())
            }),
        )
        .unwrap()
        .result();
    results.insert(file_id, res.into());

    results.into()
}
