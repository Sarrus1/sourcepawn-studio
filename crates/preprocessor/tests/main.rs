use serde::Serialize;
use vfs::FileId;

use insta::{assert_json_snapshot, assert_snapshot};

fn extend_macros(
    _macro_store: &mut MacrosMap,
    mut _path: String,
    _file_id: FileId,
    _quoted: bool,
) -> anyhow::Result<()> {
    Ok(())
}

#[derive(Debug, Default, Serialize)]
struct Range {
    start: u32,
    end: u32,
}

#[derive(Debug, Default, Serialize)]
struct ExpandedSymbol {
    range: Range,
    expanded_range: Range,
    idx: u32,
    file_id: u32,
}

impl From<&ExpandedSymbolOffset> for ExpandedSymbol {
    fn from(value: &ExpandedSymbolOffset) -> Self {
        Self {
            range: Range {
                start: value.range().start().into(),
                end: value.range().end().into(),
            },
            expanded_range: Range {
                start: value.expanded_range().start().into(),
                end: value.expanded_range().end().into(),
            },
            idx: value.idx(),
            file_id: value.file_id().0,
        }
    }
}

#[derive(Debug, Default, Serialize)]
struct PreprocessingResult_ {
    vec: Vec<(u32, u32, u32, u32)>,
    expanded_symbols: Vec<ExpandedSymbol>,
}

impl From<PreprocessingResult> for PreprocessingResult_ {
    fn from(value: PreprocessingResult) -> Self {
        Self {
            vec: value
                .source_map()
                .u_range_to_s_range_vec()
                .iter()
                .map(|e| {
                    (
                        e.0.start().into(),
                        e.0.end().into(),
                        e.1.start().into(),
                        e.1.end().into(),
                    )
                })
                .collect(),
            expanded_symbols: value
                .source_map()
                .expanded_symbols()
                .iter()
                .map(ExpandedSymbol::from)
                .collect(),
        }
    }
}

#[allow(unused_macros)]
macro_rules! assert_preproc_eq {
    ($input:expr) => {
        let res = SourcepawnPreprocessor::new(FileId::from(0), $input, &mut extend_macros)
            .preprocess_input();
        assert_snapshot!(res.preprocessed_text());
        /*
        // Disabled because redundant
        for (u_range, s_range) in res.source_map().u_range_to_s_range_vec() {
            let u_start: u32 = u_range.start().into();
            let u_end: u32 = u_range.end().into();
            let s_start: u32 = s_range.start().into();
            let s_end: u32 = s_range.end().into();
            let u_slc = u_start as usize..u_end as usize;
            let s_slc = s_start as usize..s_end as usize;
            assert_eq!(
                $input[u_slc.clone()],
                res.preprocessed_text()[s_slc.clone()],
                "{:?} does not map to {:?}: {} is different from {}",
                u_range,
                s_range,
                $input[u_slc].to_string(),
                res.preprocessed_text()[s_slc].to_string()
            )
        }
        */
        assert_json_snapshot!(PreprocessingResult_::from(res));
    };
}

use preprocessor::{ExpandedSymbolOffset, MacrosMap, PreprocessingResult, SourcepawnPreprocessor};
#[test]
fn no_preprocessor_directives() {
    let input = r#"
        int foo;
        int bar;
        "#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_defined() {
    let input = r#"#define FOO
#if defined FOO
    int foo;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_defined_complex_1() {
    let input = r#"#define FOO
#if defined FOO && defined BAR
    int foo;
    int bar;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_defined_complex_2() {
    let input = r#"#define FOO
#define BAR
#if defined FOO && defined BAR
    int foo;
    int bar;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_defined_complex_3() {
    let input = r#"#define FOO
#define BAR
#if defined FOO
    int foo;
    #if defined BAR
    int bar;
    #endif
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_defined_complex_4() {
    let input = r#"#define FOO
#if defined FOO
    int foo;
    #if defined BAZ
    int bar;
    #else
    int baz;
    #endif
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_defined_complex_5() {
    let input = r#"#define FOO
#if defined FOO
int foo;
#else
int bar;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_defined_complex_6() {
    let input = r#"#define FOO
#if defined BAR
int foo;
#else
int bar;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_defined_complex_7() {
    let input = r#"#define FOO
#if defined FOO
public void OnPluginStart()
#else
#if 1==1
public void OnPluginStart(int args)
#else
int bar;
#endif
#endif
{
    int bar;
}"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_defined_complex_8() {
    let input = r#"#define FOO 1
#if defined FOO
int foo;
#elseif FOO == 1
int bar;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_1() {
    let input = r#"#define FOO 1"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_trailing_comment_1() {
    let input = r#"#define FOO 1 /* comment */
int foo = 1;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_trailing_comment_3() {
    let input = r#"#define FOO bar /**< documentation */
#include "file.sp"
#if defined BAZ
int foo;

#endif
"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_expansion_1() {
    let input = r#"#define FOO 1
#if FOO
    int foo;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_expansion_2() {
    let input = r#"#define FOO 1
#if FOO == 1
    int foo;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_expansion_3() {
    let input = r#"#define FOO(%0,%1) %0 + %1
#define BAR(%0,%1) %0 + %1
#if FOO(1, BAR(2, 3 + 4)) == 10
#endif
"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_nested_expansion_1() {
    let input = r#"#define FOO BAR
#define BAR 1
#if FOO == 1
    int foo;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_nested_expansion_2() {
    let input = r#"#define FOO BAR + 4
#define BAR 1 + BAZ
#define BAZ 2 + 3
#if FOO == 10
    int foo;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn if_directive_nested_expansion_infinite_loop_1() {
    let input = r#"#define FOO BAR
#define BAR FOO
#if FOO == 1
    int foo;
#endif"#;

    assert_preproc_eq!(input);
}

#[test]
fn multiline_block_comment_1() {
    let input = r#"int foo;
#if false
  /*
    A
block
            comment
        */
#endif
int bar;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn multiline_block_comment_2() {
    let input = r#"int foo;
#if false /*
    A
block
            comment
        */ #endif
int bar;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn elseif_directive_expansion_1() {
    let input = r#"#define FOO 1
#if FOO == 2
int foo;
#elseif FOO == 1
int bar;
#endif
"#;

    assert_preproc_eq!(input);
}

#[test]
fn elseif_directive_expansion_2() {
    let input = r#"#define FOO 1
#if FOO == 1
int foo;
#elseif FOO == 2
int bar;
#endif
"#;

    assert_preproc_eq!(input);
}

#[test]
fn elseif_directive_expansion_3() {
    let input = r#"#define FOO 1
#if FOO == 1
    #if FOO == 1
    int foo;
    #endif
#elseif FOO == 2
int bar;
#endif
"#;

    assert_preproc_eq!(input);
}

#[test]
fn elseif_directive_expansion_4() {
    let input = r#"#define FOO 3
#if FOO == 1
int foo;
#elseif FOO == 2
int bar;
#else
int baz;
#endif
"#;

    assert_preproc_eq!(input);
}

#[test]
fn elseif_directive_expansion_5() {
    let input = r#"#define FOO 3
#if FOO == 1
#if FOO == 3
int foo;
#endif
#elseif FOO == 2
int bar;
#else
int baz;
#endif
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_undef_1() {
    let input = r#"#define FOO 1
#undef FOO
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_1() {
    let input = r#"#define FOO 1
int foo = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_2() {
    let input = r#"#define FOO "test"
char foo[64] = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_comment_1() {
    let input = r#"#define FOO 1 //comment
int foo = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_comment_2() {
    let input = r#"#define FOO 1 /* comment */
int foo = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_comment_3() {
    let input = r#"#define FOO 1 /* long\
comment */
int foo = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_comment_4() {
    let input = r#"#define FOO 1 /* long\
comment */ + 2
int foo = FOO;
int bar;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_comment_5() {
    let input = r#"#define FOO 1 /* long\
comment */ + 2 // Line comment
int foo = FOO;
int bar;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_line_continuation_1() {
    let input = r#"#define FOO "test \
expansion"
char foo[64] = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_line_continuation_2() {
    let input = r#"#define FOO "test \
expansion \
\
also"
char foo[64] = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_nested_1() {
    let input = r#"#define FOO BAR +   2
#define BAR 1
int foo = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_nested_2() {
    let input = r#"#define FOO BAR + 3
#define BAR 1 + BAZ
#define BAZ 2
int foo = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_nested_3() {
    let input = r#"#define FOO BAR + 3
#define BAR 1 + 2
int foo = FOO;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn define_expansion_nested_4() {
    let input = r#"#define FOO 1 + 2
#define BAR FOO + 3 + 4
int bar = BAR;
"#;

    assert_preproc_eq!(input);
}

#[test]
fn macro_expansion_1() {
    let input = r#"#define FOO(%0,%1) %0 + %1
int foo = FOO(1, 2);
"#;

    assert_preproc_eq!(input);
}

#[test]
fn macro_expansion_2() {
    let input = r#"#define FOO(%0) %0 %%2
int foo = FOO(2);
"#;

    assert_preproc_eq!(input);
}

#[test]
fn macro_expansion_3() {
    let input = r#"#define FOO(%0,%1) %0 + %1
#define BAR(%0,%1) %0 + %1
int foo = FOO(1, BAR(2, 3 + 4));
"#;

    assert_preproc_eq!(input);
}

#[test]
fn macro_expansion_4() {
    let input = r#"#define FOO(%1,%2) %1 + %2
int foo = FOO(1, 2);
"#;

    assert_preproc_eq!(input);
}

#[test]
fn macro_expansion_5() {
    let input = r#"#define FOO(%1) int %1
FOO(foo, bar);
"#;

    assert_preproc_eq!(input);
}

#[test]
fn macro_expansion_6() {
    let input = r#"#define GET_VALUE(%1,%2) \
    public %1 Get%2(){ \
        %1 i; \
        this.GetValue("m_" ... #%2, i); \
        return i;}
        
        GET_VALUE(void, Foo)"#;

    assert_preproc_eq!(input);
}

#[test]
fn macro_expansion_7() {
    let input = r#"#define FOO(%0,%1) %0 + %1
#define BAR(%0,%1) 1 + FOO(%0, %1)
int foo = BAR(2, 3)
"#;

    assert_preproc_eq!(input);
}

#[test]
fn macro_no_expansion_1() {
    let input = r#"#define FOO(%1) #%1
public void OnPluginStart() {
    PrintToServer(FOO /*foo*/ (foo));
}
"#;

    assert_preproc_eq!(input);
}

#[test]
fn macro_no_expansion_2() {
    let input = r#"#define FOO(%1) #%1
public void OnPluginStart() {
    PrintToServer(FOO
        (foo));
}
"#;
    assert_preproc_eq!(input);
}

#[test]
fn macro_no_expansion_3() {
    let input = r#"#define FOO(%1) #%1
public void OnPluginStart() {
    PrintToServer(FOO);
}
"#;
    assert_preproc_eq!(input);
}

#[test]
fn macro_no_expansion_4() {
    let input = r#"#define FOO(%1) %1
#define BAR(%1) %1
public void OnPluginStart() {
    PrintToServer(BAR(FOO));
}
"#;

    assert_preproc_eq!(input);
}

#[test]
fn include_directive_1() {
    let input = r#"#include <sourcemod>"#;

    assert_preproc_eq!(input);
}

#[test]
fn include_directive_2() {
    let input = r#"#include <sourcemod\
>"#;

    assert_preproc_eq!(input);
}

#[test]
fn include_directive_3() {
    let input = r#"#include <sourcemod>
"#;

    assert_preproc_eq!(input);
}

#[test]
fn stringizing_1() {
    let input = r#"#define FOO(%0) #%0
char foo[8] = FOO(foo);"#;

    assert_preproc_eq!(input);
}

#[test]
fn stringizing_2() {
    let input = r#"#define FOO(%0,%1) #%0 ... #%1
char foo[8] = FOO(foo, bar);"#;

    assert_preproc_eq!(input);
}

#[test]
fn stringizing_3() {
    let input = r#"#define FOO(%0) #%0
char foo[8] = FOO(foo , bar);"#;

    assert_preproc_eq!(input);
}

#[test]
fn stringizing_4() {
    let input = r#"#define DISPOSE_MEMBER(%1) \
    Handle m_h%1; \
    if(this.GetValue("m_" ... #%1, m_h%1)){ \
        delete m_h%1;}
void foo(){
    DISPOSE_MEMBER(Foo)
}"#;

    assert_preproc_eq!(input);
}
