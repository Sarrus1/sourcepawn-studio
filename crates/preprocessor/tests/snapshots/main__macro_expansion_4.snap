---
source: crates/preprocessor/tests/main.rs
assertion_line: 678
expression: "SourcepawnPreprocessor::new(FileId::from(0), input,\n                &mut extend_macros).preprocess_input().preprocessed_text().as_ref()"
---
#define FOO(%1,%2) %1 + %2
int foo = 1 + 2;

